### PR TITLE
Pv2 170 small employers act2 tt 1 2 3

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.PaymentsDue.PaymentsDueService/Handlers/ApprenticeshipContractType2PayableEarningEventHandler.cs [hunk 1]
```diff
@@ -1,73 +1,72 @@
-﻿using System;
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.PaymentsDue.PaymentsDueService/Handlers/ApprenticeshipContractType2PayableEarningEventHandler.cs]: System.Collections.Generic is unused

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/OnProgrammePaymentDue.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using SFA.DAS.Payments.Model.Core.OnProgramme;
+
+namespace SFA.DAS.Payments.PaymentsDue.AcceptanceTests.Data
+{
+    public class OnProgrammePaymentDue
+    {
+        public string PriceEpisodeIdentifier { get; set; }
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/OnProgrammePaymentDue.cs]: It should be used. Fixed.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/ContractTypeEarning.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace SFA.DAS.Payments.PaymentsDue.AcceptanceTests.Data
+{
+    public class ContractTypeEarning
+    {
+        public string PriceEpisodeIdentifier { get; set; }
+
+        public byte Delivery_Period { get; set; }
+
+        public OnProgrammeEarningType Type => TransactionType.Contains("Learning") ?
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/ContractTypeEarning.cs]: Yes, they are very similar. Refactored.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/OnProgrammePaymentDue.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        public  byte Delivery_Period { get; set; }
+        public decimal Amount { get; set; }
+        public OnProgrammeEarningType Type => TransactionType.Contains("Learning") ?
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Data/OnProgrammePaymentDue.cs]: Yes, they are very similar. Refactored.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Handlers/ApprenticeshipContractType2PaymentDueEventHandler.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace SFA.DAS.Payments.PaymentsDue.AcceptanceTests.Handlers
+{
+    public class ApprenticeshipContractType2PaymentDueEventHandler : IHandleMessages<ApprenticeshipContractType2PaymentDueEvent>
+    {
+        public static ConcurrentBag<ApprenticeshipContractType2PaymentDueEvent> ReceivedEvents { get; } = new ConcurrentBag<ApprenticeshipContractType2PaymentDueEvent>();
+
+        public Task Handle(ApprenticeshipContractType2PaymentDueEvent message, IMessageHandlerContext context)
+        {
+            Console.WriteLine(message.ToJson());
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Handlers/ApprenticeshipContractType2PaymentDueEventHandler.cs]: I don't think this should be there at all. Removed.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/EarningSteps.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/EarningSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/EarningSteps.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        [Given(@"the SFA contribution percentage is (.*)%")]
+        [When(@"the SFA contribution percentage changes to (.*)%")]
+        public void SFAContributionPercentageIs(decimal sfaContributionPercentage)
+        {
+            Console.WriteLine($"Got sfa contribution percentage: {sfaContributionPercentage}");
+            SfaContributionPercentage = sfaContributionPercentage;
+        }
+
+
+        //[Given(@"the following contract type (.*) on programme earnings for periods (.*)-(.*) are provided in the latest ILR for the academic year (.*):")]
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/EarningSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
+﻿using SFA.DAS.Payments.AcceptanceTests.Core;
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    public class LearnerSteps
+    {
+        private LearnRefNumberGenerator learnRefNumberGenerator;
+
+        public LearnerSteps(ScenarioContext context)
+        {
+            this.learnRefNumberGenerator = new LearnRefNumberGenerator(context);
+        }
+
+        //[BeforeScenario]
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueBootstrapper.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+﻿using Autofac;
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueBootstrapper.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            var expectedPaymentsEvents = table.CreateSet<OnProgrammePaymentDue>().ToList();
+
+            this.WaitForIt(() =>
+            {
+                return expectedPaymentsEvents.All(payment => act == 1 ? Act1Matcher(payment) : Act2Matcher(payment));
+            }, "Failed to find all the payment due events");
+        }
+
+        private bool Act1Matcher(OnProgrammePaymentDue expectedEvent)
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs]: Refactored to remove the need for abbreviation.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            //yesScenarioCtx[$"ContractType{this.ContractType}OnProgrammeEarningsLearning"] = earning;
+        }
+
+        private List<ApprenticeshipContractType2EarningEvent> Act2EarningEvents
+        {
+            get => Get<List<ApprenticeshipContractType2EarningEvent>>();
+            set => Set(value);
+        }
+
+        protected PaymentsDueSteps(ScenarioContext scenarioContext) : base(scenarioContext)
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                                {
+                                    Period = new CalendarPeriod(this.CollectionYear, e.Delivery_Period),
+                                    Amount = e.Amount,
+                                    PriceEpisodeIdentifier = e.PriceEpisodeIdentifier
+                                }).ToList())
+                            }).ToArray())
+                }
+            };
+
+            //yesScenarioCtx[$"ContractType{this.ContractType}OnProgrammeEarningsLearning"] = earning;
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/PaymentsDueSteps.cs]: done

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/ProcessingPeriodSteps.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        public void GivenTheCurrentProcessingPeriodIs(byte period)
+        {
+            CollectionPeriod = period;
+        }
+
+        [Given(@"the payments are for the current collection year")]
+        public void GivenThePaymentsAreForTheCurrentCollectionYear()
+        {
+            var year = DateTime.Today.Year - 2000;
+            CollectionYear = DateTime.Today.Month < 9 ? $"{year - 1}{year}" : $"{year}{year + 1}";
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/ProcessingPeriodSteps.cs]: implementation moved to StepsBase. Given binding cannot be moved.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/SmallEmployerPaymentsSteps.cs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        
+        [Given(@"the required payments component generates the following contract type (.*) payable earnings:")]
+        public void GivenTheRequiredPaymentsComponentGeneratesTheFollowingContractTypePayableEarnings(int p0, Table table)
+        {
+            ScenarioContext.Current.Pending();
+        }
+        
+        [When(@"required payments event is received")]
+        public void WhenRequiredPaymentsEventIsReceived()
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/SmallEmployerPaymentsSteps.cs]: factored out

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/SmallEmployerPaymentsSteps.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            ScenarioContext.Current.Pending();
+        }
+        
+        [Then(@"the payment source component will generate the following contract type (.*) coinvested payments:")]
+        public void ThenThePaymentSourceComponentWillGenerateTheFollowingContractTypeCoinvestedPayments(int p0, Table table)
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/SmallEmployerPaymentsSteps.cs]: factored out

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
+﻿using SFA.DAS.Payments.AcceptanceTests.Core;
+using SFA.DAS.Payments.AcceptanceTests.Core.Automation;
+using TechTalk.SpecFlow;
+
+namespace SFA.DAS.Payments.PaymentsDue.AcceptanceTests.Steps
+{
+    [Binding]
+    public class LearnerSteps
+    {
+        private LearnRefNumberGenerator learnRefNumberGenerator;
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs]: it should be removed. done.

File: src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs [hunk 18]
```diff
@@ -0,0 +1,28 @@
+﻿using SFA.DAS.Payments.AcceptanceTests.Core;
+using SFA.DAS.Payments.AcceptanceTests.Core.Automation;
+using TechTalk.SpecFlow;
+
+namespace SFA.DAS.Payments.PaymentsDue.AcceptanceTests.Steps
+{
+    [Binding]
+    public class LearnerSteps
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.AcceptanceTests/Steps/LearnerSteps.cs]: yes

File: src/SFA.DAS.Payments.PaymentsDue.PaymentsDueService/Handlers/ApprenticeshipContractType2PayableEarningEventHandler.cs (hunk 1 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.PaymentsDue.PaymentsDueService/Handlers/ApprenticeshipContractType2PayableEarningEventHandler.cs]: done

**Pull request merged**